olive/passes/pytorch/lora.py

```diff
@@ -191,26 +195,162 @@ def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigPa
+                model.enable_input_require_grads()
+            else:
+
+                def make_inputs_require_grad(module_, input_, output_):
+                    output_.requires_grad_(True)
+
+                model.get_input_embeddings().register_forward_hook(make_inputs_require_grad)
+
+            # enable gradient checkpointing for memory efficiency
+            model.gradient_checkpointing_enable()
+
+        return model
+
+    def enable_lora(
+        self,
+        model: PreTrainedModel,
+        tokenizer: PreTrainedTokenizer,
+        task: str,
+        config: ConfigBase,
+        target_modules: Optional[List[str]],
+    ) -> torch.nn.Module:
+        """Run LoRA fine-tuning on a Hugging Face PyTorch model."""
+        from peft import LoraConfig, get_peft_model
+        from peft.tuners.lora import LoraLayer
+
+        logger.debug("Adding LoRA modules")
+        # if there is no pad token, add to tokenizer and model
+        # TODO(jambayk): Do this in a better way since the embedding size might become unoptimal
+        # (not a multiple of 64, etc) perhaps use eos_token as pad_token, but need to ensure the actual eos_token
+        # at the end of the sequence is not masked (both in attention mask and loss calculation)
+        if not tokenizer.pad_token_id:
+            self.smart_tokenizer_and_embedding_resize(
+                special_tokens_dict={"pad_token": DEFAULT_PAD_TOKEN}, tokenizer=tokenizer, model=model
+            )
+
+        model = self.prepare_model_for_lora_finetuning(model, config.training_args.gradient_checkpointing)
+
+        # set model_parallel and is_parallelizable to True
+        # we are using "auto" device_map, so model_parallel is True or doing DDP
+        # don't want the trainer to do Data Parallel
+        setattr(model, "model_parallel", True)
+        setattr(model, "is_parallelizable", True)
+
+        logger.debug(
+            f"The number of trainable parameters in the original model: {self.count_trainable_parameters(model)}"
+        )
+        peft_task_type = get_peft_task_type_from_task(task, fail_on_not_found=True)
+        lora_config = LoraConfig(
+            r=config.lora_r,
+            lora_alpha=config.lora_alpha,
+            lora_dropout=config.lora_dropout,
+            target_modules=target_modules,
+            bias=config.bias,
+            task_type=peft_task_type,
+            modules_to_save=config.modules_to_save,
+        )
+
+        lora_model = get_peft_model(model, lora_config)
+        logger.debug(
+            f"The number of trainable parameters in the LoRA model: {self.count_trainable_parameters(lora_model)}"
+        )
+
+        # cast lora modules to model's dtype, should be same as torch_dtype
+        for module in lora_model.modules():
+            if isinstance(module, LoraLayer):
+                module.to(lora_model.dtype)
+
+        return lora_model
+
+    def train_and_save_new_model(
```

Comment:
I don't see the device is used in training?

Reply:
Do you mean the gpu? It is enabled by `device_map="auto"` during model loading.